File: BddPipe/BddPipe/StringExtensions.cs
using System;
using static BddPipe.F;

namespace BddPipe
{
    internal static class StringExtensions
    {
        private static bool StartsWithIgnoreCase(this in Some<string> text, in Some<string> prefix) =>
            text.Value.IndexOf(prefix.Value, StringComparison.InvariantCultureIgnoreCase) == 0;

        public static Some<string> WithPrefix(this in Option<string> text, Some<string> prefix) =>
            text.Match(
                txt =>
                {
                    Some<string> someText = txt;

                    var result = someText.StartsWithIgnoreCase(prefix)
                        ? someText
                        : new Some<string>($"{prefix} {txt}".TrimEnd());

                    return result;
                },
                () => prefix
            );

        public static Option<string> NoneIfWhiteSpace(this in Option<string> text) =>
            text.Match<Option<string>>(
                txt =>
                {
                    if (string.IsNullOrWhiteSpace(txt))
                    {
                        return None;
                    }

                    return txt;
                },
                () => None
            );
    }
}

File: BddPipe/BddPipe.FSharp/StringExtensions.fs
namespace BddPipe.FSharp

open System

module internal StringExtensions =
    let private startsWithIgnoreCase (text: string) prefix =
        text.IndexOf(prefix, StringComparison.InvariantCultureIgnoreCase) = 0

    let withPrefix prefix text =
        match text with
            | Some txt ->
                if (startsWithIgnoreCase txt prefix) then
                    txt
                else
                    (sprintf "%s %s" prefix txt).TrimEnd()
            | None -> prefix
